[#120] Ported sys:setup:change-version command

robbieaverill opened the pull request:
Resolved #120.

Changes:

* Ported sys:setup:change-version command
* Updated references to M1 framework for M2 in abstract command
* Added unit tests

Notes:

* I've removed the setup argument as it appears that Magento 2 has removed the granularity of different setup resources.
* ~~I've removed the inject() method use in favour of mockable getters (although it pained me to do it)~~ Update: left in place, using getters as well for mocking
* ~~The tests that mock the command classes need to be run in isolation, otherwise the following error is thrown: `PHP Fatal error:  Call to undefined method Magento\Framework\Module\ModuleList::getNames() in [.../]src/vendor/magento/framework/Module/DependencyChecker.php on line 50`~~ Not required any longer.

robbieaverill commented:
Well, I'm obviously doing something wrong. Annoying that this doesn't happen locally. Will take another look tomorrow.

ktomk commented:
Hi @robbieaverill, I could only quickly peek into the changes and errors, and I do not have much of an idea. For what is reported on Travis in the test-suite, the *"Constant ... already defined"* errors, is normally a sign that the file that is defining the constant (one constant-name given is `BP` and the file is `app/autoload.php`) is included more than once. Perhaps looking what triggers the inclusion sheds more light. The Travis build always installs from scratch and this could make it different to you local tests.

robbieaverill commented:
@ktomk I was trying to find an example of a unit test that mocks the command but couldn't see one immediately - it's likely a problem with how I'm doing that. I don't believe it's the wrong way to do it, but perhaps I've missed something in the process like a constructor. Will keep looking...

ktomk commented:
@robbieaverill: I could take another look over the changes. I have left some comments, please see if they make sense to you. Was a bit in a hurry so excuse if my tone sounds harsh, it's absolutely not meant that way. I try to think about your comments a bit more, esp. regarding the questions you raise about the testing. Thanks again for all your hard work!

robbieaverill commented:
Hi @ktomk - thanks for the code review and the suggestions, I've made the following changes and updated the branch:

* `getMagentoModuleList()` now calls the ModuleList's `getAll()` method internally and returns an array
* Update unit tests to reflect this
* Added the previous module version to the `change-version` command output

Cheers,
Robbie

ktomk commented:
Hey @robbieaverill, nice update. I did review the changes a bit, looks good. What I don't understand - but most likely I'm overlooking something  - is how in live-mode the injection of the modulelist and the setup resource is done? The inject method is only called in tests. So I'm wondering a bit how this works.

robbieaverill commented:
Tom- good question :-) i used it because it was there already, haven't
looked at how it works haha

On Saturday, 2 January 2016, Tom Klingenberg <notifications@github.com>
wrote:

> Hey @robbieaverill <https://github.com/robbieaverill>, nice update. I did
> review the changes a bit, looks good. What I don't understand - but most
> likely I'm overlooking something - is how in live-mode the injection of the
> modulelist and the setup resource is done? The inject method is only called
> in tests. So I'm wondering a bit how this works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/netz98/n98-magerun2/pull/167#issuecomment-168339505>.
>


-- 
*Robbie Averill*
LA PETITE MANOUCHE - NZ Gypsy Jazz Guitar Duo
Phone: +64 27 321 4758
contact@lapetitemanouche.com
www.lapetitemanouche.com
